Update Semantics doc about changes in GCSFuse behavior after Write Failure Improvements (#2877)
GoogleCloudPlatform/gcsfuse: docs/semantics.md

**Concurrency**

> Multiple readers can access the same or different objects within a bucket without issue. Likewise, multiple writers modify different objects in the same bucket simultaneously without any issue. Even when multiple writers attempt to replace the same file from the same mount point, the last write wins similar to a native filesystem.

Reviewer: The last write wins is a little confusing. Can we instead write - concurrent writes to the same gcs object are supported from the same mount and behave similar to native file system.

Author: Yeah, sounds good.

> However, when different mounts try to write to the same object, the flush from first mount wins. Other mounts that have not updated their local file descriptors after the object is updated will encounter a ```syscall.ESTALE``` error when attempting to save their edits due to precondition checks. Therefore, to ensure data integrity it is strongly recommended that multiple sources do not modify the same object.

Reviewer: Nit:
```suggestion
However, when different mounts try to write to the same object, the flush from first mount wins. Other mounts that have not updated their local file descriptors after the object is modified will encounter a ```syscall.ESTALE``` error when attempting to save their edits due to precondition checks. Therefore, to ensure data is consistently written, it is strongly recommended that multiple sources do not modify the same object.
```

**Concurrency**

> Multiple readers can access the same or different objects from the same bucket without issue. Multiple writers can also write to different objects in the same bucket without issue. However, there is no concurrency control for multiple writers to the same file. When multiple writers try to replace a file from the same machine, the last write wins and all previous writes are lost without any user notification of the overwrite. However, in case of different machines the flush from first machine wins and the other machines that had the file open before the first machine synced its changes will get a ```syscall.ESTALE``` error when they try to save their own edits - there is no merging or version control. Therefore, for data integrity it is recommended that multiple sources do not modify the same object.

Author: Updated as per discussion with @charith87

- Machine A opens a file and writes then successfully closes or syncs it, and the file was not concurrently unlinked from the point of view of A. Machine B then opens the file after machine A finishes closing or syncing. Machine B will observe a version of the file at least as new as the one created by machine A.
- Machine A and B both open the same file, which contains the text ‘ABC’. Machine A modifies the file to ‘ABC-123’ and closes/syncs the file which gets written back to Cloud Storage. After, Machine B, which still has the file open, instead modifies the file to ‘ABC-XYZ’, and saves and closes the file. As the last writer wins, the current state of the file will read ‘ABC-XYZ’.

**Stale File Handle Errors**

> To prevent data corruption and ensure consistency, Cloud Storage FUSE actively detects and handles situations that could lead to stale file handles. This results in a ```syscall.ESTALE``` error under the following circumstances:

Reviewer: I think data corruption is not the correct term here. 
```suggestion
To ensure consistency, Cloud Storage FUSE returns a ```syscall.ESTALE``` error when an application tries to access stale data. This can occur in the following circumstances:
```

Author: Agreed

- **Concurrent Writes**: When multiple mounts have the same file open for writing, and one mount modifies and syncs the file, other mounts with open file handles will encounter this error when the writer application attempts to sync or close the file.

> - **Read During Modification**: If one mount is reading a file while another mount modifies and syncs the file, the reader application will encounter this error.
> - **File Deletion or Renaming During Read**: If a mount is reading a file and the file is deleted or renamed by the same or a different mount, the reader application will encounter this error.

Reviewer: Delete and Rename also comes under file modification. I think we can combine it this way.
```suggestion
- **Read During Modification**:  If an application is reading a file through a GCSFuse mount, and the same object is modified on GCS (by deleting, renaming, or changing its content or metadata), the GCSFuse reader will encounter this error. This is because GCSFuse detects that the file it was accessing has changed.
```

> - **File Renaming During Write**: If a mount is writing to a file and the file is renamed by the same or a different mount, the writer application will encounter this error when attempting to sync or close the file.

Reviewer: ```suggestion
- **File Renaming During Write**: If an application is writing to a file through a GCSFuse mount, and the same object is renamed on Google Cloud Storage (via same or different GCSFuse mount or through another interface), the writer will encounter this error when syncing or closing the file.
```

> - **File Deletion During Write**: If a mount is writing to a file and the file is deleted by the same or a different mount, the writer application will encounter this error when attempting to sync or close the file. However, if the file is deleted by the same mount, the writer application will not encounter any error.

Reviewer: ```suggestion
- **File Deletion During Write**:  If an application is writing to a file through a GCSFuse mount, and the same object is deleted on Google Cloud Storage (via different GCSFuse mount or through another interface), the writer will encounter this error when syncing or closing the file.
```

- Machine A opens a file and writes then successfully closes or syncs it, and the file was not concurrently unlinked from the point of view of A. Machine B then opens the file after machine A finishes closing or syncing. Machine B will observe a version of the file at least as new as the one created by machine A.
- Machine A and B both open the same file, which contains the text ‘ABC’. Machine A modifies the file to ‘ABC-123’ and closes/syncs the file which gets written back to Cloud Storage. After, Machine B, which still has the file open, instead modifies the file to ‘ABC-XYZ’, and saves and closes the file. As the last writer wins, the current state of the file will read ‘ABC-XYZ’.

**Stale File Handle Errors**

To ensure consistency, Cloud Storage FUSE returns a ```syscall.ESTALE``` error when an application tries to access stale data. This can occur in the following circumstances:

> - **Concurrent Writes**: When multiple mounts have the same file open for writing, and one mount modifies and syncs the file, other mounts with open file handles will encounter this error when the writer application attempts to sync or close the file.

Reviewer: Nit:
```suggestion
- **Concurrent Writes**: When multiple mounts have the same file open for writing, and one mount modifies and syncs the file, other mounts with open file descriptors will encounter this error when attempting to sync or close the file.
```

- **Read During Modification**:  If an application is reading a file through a GCSFuse mount, and the same object is modified on GCS (by deleting, renaming, or changing its content or metadata), the GCSFuse reader will encounter this error. This is because GCSFuse detects that the file it was accessing has changed.

> - **File Renaming During Write**: If an application is writing to a file through a GCSFuse mount, and the same object is renamed on Google Cloud Storage (via same or different GCSFuse mount or through another interface), the writer will encounter this error when syncing or closing the file.

Reviewer: Nit: Can you change If -> When in all the bullet points for consisitency
